dockerfiles/enclave.Dockerfile

```diff
@@ -1,31 +1,39 @@
-FROM ghcr.io/edgelesssys/ego-dev:latest
+FROM ghcr.io/edgelesssys/ego-dev:latest AS build-base
 # on the container:
 #   /home/obscuro/data       contains working files for the enclave
 #   /home/obscuro/go-obscuro contains the src
 #
 
 # setup container data structure
-RUN mkdir -p /home/obscuro/data && mkdir -p /home/obscuro/go-obscuro
+RUN mkdir -p /home/obscuro/go-obscuro
 
 # Ensures container layer caching when dependencies are not changed
 WORKDIR /home/obscuro/go-obscuro
 COPY go.mod .
 COPY go.sum .
-RUN go mod download
+RUN ego-go mod download
 
-# COPY the source code as the last step
+
+# Trigger new build stage for compiling the enclave
+FROM build-base as build-enclave
 COPY . .
 
-# build binary
 WORKDIR /home/obscuro/go-obscuro/go/enclave/main
-RUN ego-go build
+
+# Build the enclave using the cross image build cache.
+RUN --mount=type=cache,target=/root/.cache/go-build \
```

Comment:
I'm weary that this might break (like builderx did on the faucet) and we had to strip it out for a simpler docker build.

What if it was narrowed to the debug images given that's mainly for local speed improvement ?

Reply:
This shouldn't break (or in fairness do anything) in the CI process as the runner will always be a fresh one. Also the main speed improvement is coming from the enclave building which takes 300~ seconds normally (signing is 30 seconds)

There is a standard github action for docker build & push that has configurations to allow cross run sharing of the cache, but not much point moving to them.

Reply:
OOC, why is `--mount=type=cache,target=/root/.cache/go-build  ego-go build` faster ?

Reply:
It shares the go compiler build cache across the layer rebuilds of the image and thus allows the go compiler to work incrementally rather than recompile everything from scratch everytime. This is a speedup only for the local builds obviously as there isn't any cache on the CI. The speedup on the CI is because of the docker builder being able to launch downloads in parallel. (the CI l2 image build drops from 9 mintues to 6, not that big of a difference)

Reply:
Cool, thanks for the info :)